From 27a9fe59674c5349c295eee9ad914153a684ab24 Mon Sep 17 00:00:00 2001
Message-Id: <27a9fe59674c5349c295eee9ad914153a684ab24.1646404612.git.lorenzo@kernel.org>
In-Reply-To: <f39765369a03ef8f03c32cba4b2459e13eaf321a.1646404612.git.lorenzo@kernel.org>
References: <f39765369a03ef8f03c32cba4b2459e13eaf321a.1646404612.git.lorenzo@kernel.org>
From: Lorenzo Bianconi <lorenzo@kernel.org>
Date: Fri, 4 Mar 2022 13:27:39 +0100
Subject: [PATCH 4/8] DFS: rely on channel_type in dfs_downgrade_bandwidth
 signature

Add the capability to specify all 3 channel type possibilities in
dfs_downgrade_bandwidth routine.
This is a preliminary change to introduce radar/CAC background detection
support.

Signed-off-by: Lorenzo Bianconi <lorenzo@kernel.org>
---
 src/ap/dfs.c | 31 ++++++++++++++-----------------
 1 file changed, 14 insertions(+), 17 deletions(-)

--- a/src/ap/dfs.c
+++ b/src/ap/dfs.c
@@ -987,7 +987,7 @@ int hostapd_dfs_pre_cac_expired(struct h
 static struct hostapd_channel_data *
 dfs_downgrade_bandwidth(struct hostapd_iface *iface, int *secondary_channel,
 			u8 *oper_centr_freq_seg0_idx,
-			u8 *oper_centr_freq_seg1_idx, int *skip_radar)
+			u8 *oper_centr_freq_seg1_idx, int *channel_type)
 {
 	struct hostapd_channel_data *channel;
 
@@ -995,23 +995,22 @@ dfs_downgrade_bandwidth(struct hostapd_i
 		channel = dfs_get_valid_channel(iface, secondary_channel,
 						oper_centr_freq_seg0_idx,
 						oper_centr_freq_seg1_idx,
-						*skip_radar ? DFS_AVAILABLE :
-						DFS_ANY_CHANNEL);
+						*channel_type);
 		if (channel) {
 			wpa_printf(MSG_DEBUG, "DFS: Selected channel: %d",
 				   channel->chan);
 			return channel;
 		}
 
-		if (*skip_radar) {
-			*skip_radar = 0;
+		if (*channel_type) {
+			*channel_type = DFS_ANY_CHANNEL;
 		} else {
 			int oper_chwidth;
 
 			oper_chwidth = hostapd_get_oper_chwidth(iface->conf);
 			if (oper_chwidth == CHANWIDTH_USE_HT)
 				break;
-			*skip_radar = 1;
+			*channel_type = DFS_AVAILABLE;
 			hostapd_set_oper_chwidth(iface->conf, oper_chwidth - 1);
 		}
 	}
@@ -1029,7 +1028,7 @@ static int hostapd_dfs_start_channel_swi
 	int secondary_channel;
 	u8 oper_centr_freq_seg0_idx = 0;
 	u8 oper_centr_freq_seg1_idx = 0;
-	int skip_radar = 0;
+	int channel_type = DFS_ANY_CHANNEL;
 	int err = 1;
 
 	/* Radar detected during active CAC */
@@ -1037,14 +1036,13 @@ static int hostapd_dfs_start_channel_swi
 	channel = dfs_get_valid_channel(iface, &secondary_channel,
 					&oper_centr_freq_seg0_idx,
 					&oper_centr_freq_seg1_idx,
-					skip_radar ? DFS_AVAILABLE :
-					DFS_ANY_CHANNEL);
+					channel_type);
 
 	if (!channel) {
 		channel = dfs_downgrade_bandwidth(iface, &secondary_channel,
 						  &oper_centr_freq_seg0_idx,
 						  &oper_centr_freq_seg1_idx,
-						  &skip_radar);
+						  &channel_type);
 		if (!channel) {
 			wpa_printf(MSG_ERROR, "No valid channel available");
 			return err;
@@ -1078,7 +1076,7 @@ static int hostapd_dfs_start_channel_swi
 	u8 oper_centr_freq_seg0_idx;
 	u8 oper_centr_freq_seg1_idx;
 	u8 new_vht_oper_chwidth;
-	int skip_radar = 1;
+	int channel_type = DFS_AVAILABLE;
 	struct csa_settings csa_settings;
 	unsigned int i;
 	int err = 1;
@@ -1103,14 +1101,13 @@ static int hostapd_dfs_start_channel_swi
 	 * uniform spreading.
 	 */
 	if (iface->dfs_domain == HOSTAPD_DFS_REGION_ETSI)
-		skip_radar = 0;
+		channel_type = DFS_ANY_CHANNEL;
 
 	/* Perform channel switch/CSA */
 	channel = dfs_get_valid_channel(iface, &secondary_channel,
 					&oper_centr_freq_seg0_idx,
 					&oper_centr_freq_seg1_idx,
-					skip_radar ? DFS_AVAILABLE :
-					DFS_ANY_CHANNEL);
+					channel_type);
 
 	if (!channel) {
 		/*
@@ -1118,11 +1115,11 @@ static int hostapd_dfs_start_channel_swi
 		 * there is another channel where we can switch even if it
 		 * requires to perform a CAC first.
 		 */
-		skip_radar = 0;
+		channel_type = DFS_ANY_CHANNEL;
 		channel = dfs_downgrade_bandwidth(iface, &secondary_channel,
 						  &oper_centr_freq_seg0_idx,
 						  &oper_centr_freq_seg1_idx,
-						  &skip_radar);
+						  &channel_type);
 		if (!channel) {
 			/*
 			 * Toggle interface state to enter DFS state
@@ -1133,7 +1130,7 @@ static int hostapd_dfs_start_channel_swi
 			return 0;
 		}
 
-		if (!skip_radar) {
+		if (channel_type == DFS_ANY_CHANNEL) {
 			iface->freq = channel->freq;
 			iface->conf->channel = channel->chan;
 			iface->conf->secondary_channel = secondary_channel;
